src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java

```diff
@@ -26,50 +28,106 @@
-                return true;
+                try {
+                    setDisplayName(build, listener);
+	            return true;                   
+                } catch (IllegalArgumentException e){
+                    listener.getLogger().println(e.getMessage());   
+                }
+                return false;
             }
-        };
+         };
     }
-
-    private void setDisplayName(AbstractBuild build, BuildListener listener) throws IOException, InterruptedException {
-        try {
-            build.setDisplayName(TokenMacro.expandAll(build, listener, template));
-        } catch (MacroEvaluationException e) {
+    private void setDisplayName(AbstractBuild build, BuildListener listener) throws IOException, InterruptedException, IllegalArgumentException {
+    	try{
+            switch(scenario) {
+            case 0:         
+                build.setDisplayName(TokenMacro.expandAll(build, listener, matrixTemplate));
+                return;
+            case 1:
+                if(started) {	
+                    build.setDisplayName(TokenMacro.expandAll(build, listener, matrixTemplate));
+                }else {
+                    build.setDisplayName(TokenMacro.expandAll(build, listener, template));
+                }
+                return;
+            case 2:
+                build.setDisplayName(TokenMacro.expandAll(build, listener, template));
+                return;
+            case -1:
+                listener.getLogger().println("No valid values set for build names!");
+            }
+    	} catch (MacroEvaluationException e) {
             listener.getLogger().println(e.getMessage());
-        }
+    	}  
     }
 
     public MatrixAggregator createAggregator(MatrixBuild build, Launcher launcher, BuildListener listener) {
         return new MatrixAggregator(build,launcher,listener) {
             @Override
             public boolean startBuild() throws InterruptedException, IOException {
-                setDisplayName(build,listener);
-                return super.startBuild();
+                try {
+                    setDisplayName(build,listener);
+                    started = true;
+	            return super.startBuild();
+                } catch (IllegalArgumentException e){
+                    listener.getLogger().println(e.getMessage());   
+                }
+                return false;
             }
 
             @Override
             public boolean endBuild() throws InterruptedException, IOException {
-                setDisplayName(build,listener);
-                return super.endBuild();
+                try{
+                    setDisplayName(build,listener);
+                    return super.endBuild();
+                } catch (IllegalArgumentException e){
+                    listener.getLogger().println(e.getMessage());   
+                }
+                return false;
             }
         };
     }
+   
 
-    @Extension
+    @Extension(optional = true)
```

Comment:
As noted above, I do not think it is feasible for the dependency to be in any way optional currently. Would really require a new API in `matrix-project`. Anyway optional dependencies are always a headache; prefer to just keep it as a hard dep.